### ISSUE TITLE
OrchestrationTrigger: Run On Startup?

### ISSUE THREAD

**Issue by bendursley** [Azure/azure-functions-durable-extension#1536]
I'd like to be able setup some eternal orchestrations without first having to call the function, similar to how the TimerTrigger works.

I'm looking at options for some cleanup jobs which I'd like to complete and then have x time before they run again (similar to the docs example for external orcehstration).  That pattern seems to work quite well with what I'd like to do.

However, I'd like to avoid having to send a client request to the function first (HTTP Request, service bus message, etc), I'd just like it start up straight away.

**Comment by olitomlinson** [comment]
@bendursley could you use a TimerTrigger to invoke your clean up task on a desired schedule?

**Comment by bendursley** [comment]
@olitomlinson Yeah, that's my current option (and what I think I'll implement).  The eternal orchestration pattern is neater, but it needs a kick start, which is a shame.

**Comment by olitomlinson** [comment]
@bendursley Hmmmm! Are you looking at the Eternal Orchestration pattern because your clean-up task needs to maintain some form of context/state/history between invocations?

**Comment by bendursley** [comment]
@olitomlinson Was intending to run a number of jobs which clean up some data, which could take some time to run, so the option to let it finish, then add y minutes upon completion seemed a neater idea than just run every y minutes.

**Comment by olitomlinson** [comment]
@bendursley Ahh okay I get you! So you would create a timer just before calling `continueAsNew` and the timer would represent your delay until the next clean-up run? 

I'll have a think to see if there are any other patterns that come to mind that might help.

**Comment by bendursley** [comment]
@olitomlinson Yup, that was the intention, the timer would be the delay.  `CreateTimer` then `ContinueAsNew` 👍 

Thanks!

**Comment by cgillum** [comment]
I think this sounds like an interesting feature. One question would be how to handle failures. Should the orchestration automatically restart itself if there is an unhandled exception?

**Comment by bendursley** [comment]
@cgillum I think that would make sense to restart itself in x time, yes.  So long as it the error was logged

**Comment by tmenier** [comment]
Glad this is being discussed. I have the same need - fire off an "eternal" orchestration on startup. I my mind, I'm wondering if maybe there are actually 2 distinct ideas that could bring this (and other scenarios) to fruition:

1. A "run once at startup" trigger for regular Azure Functions. It would basically work like a `TimerTrigger` with `runOnStartup=true` but no actual timer functionality. I have a case where I'd like to send off an alert when a deployment happens and this would work great for that. I understand this should be suggested in a different repo though.

2. With Durable Functions specifically, sometimes trigger > regular function > orchestrator function just feels like more indirection than should be necessary. Even if it were just an abstraction that makes it _appear_ that an orchestration function is using one of the standard triggers directly, it seems like it would simplify app code in a lot of cases.

**Comment by sebastianburckhardt** [comment]
One thing to be careful about is the meaning of "on startup" which is a bit fuzzy as the state of durable functions (i.e. the taskhub state) outlives startup and shutdown cycles. For example, would a scale-from-zero situation count as a startup? 

Another interpretation is to specify that an orchestrator should be executed once in a fresh taskhub; essentially we could create the start message together with the taskhub.

**Comment by cgillum** [comment]
Indeed, "on startup" is a little fuzzy in Azure Functions since there is no concept of "startup" at the app level - rather, each individual function app instance has its own startup routine. I think the way to implement this would be to checks to see if the instance has already started each time a worker instance starts up (and start the orchestration if it's not already started). There is a question too, however, about what to do if the instance is in a failed, completed, or terminated state.

I think the problem with creating a start message together with the task hub is that a startup orchestration could be added to the app sometime after the task hub is initially created. Task hub creation only happens the first time the app is _ever_ started.

**Comment by Bartolomeus-649** [comment]
I think the correct way to do this is to introduce the **conceptual concept: "Background Job"**.
>  _And where "conceptual concept" is a two layer abstraction:_
> 1. _Concept "Background Job"
>     The concept "Background Job" here means something that not directly part of the "Business end" of a solution/system/application, but is still needed in order for everything to run smoothly and work as intended.
>     The term "Background Job" is already used in many places and when using it an experienced architect/developer immediately understand what it is and what it can and should be used for. This could both be a good thing or a bad thing, depending on how well it matches their expectation. Picking another term/name might be good if "background Job" differs in functionality/features from what is expected from a tradition background job._
> 2. _Conceptual concept "Background Job"
>    This doesn't have to be anything at all like what people associate with a background job. Background jobs are after all not the goal, but a means to achieve a goal. The important thing is to achieve the goal, not to create yet another way of creating a background job.
>    So, the important parts are the features listed below, and not at all the "Background Job"-part._ 

### This "conceptual concept" has these "features" built in:
_This list include way more than what's initially needed, the "extra" things are just here to indicate the long term direction "Background Job" should have_
* Stateless
  No state is carried over between separate runs/iterations
  
* Singleton
  The can never be more than one running at any time
  
* Always active
    There is always at least one conceptual instance running. It does not actually have to be running, it could, for example, be in a state where it waits for something to happen, like a specific point in time (I.e. 2021-10-13 22:00:00")
    
* Exception handling setting
    * Backoff strategy
        Settings which ensures there will be be no "rapid failing" scenarios where the "Background Job" starts and then immediately fail, and since the "Background Job" should always be running, the whole thing repeats in rapid successions.
            
    * Condition for when to consider the the function as "non-functional", and "pause" the execution of the "Background Job". Un-pausing require some kind of explicit action and/or event to occur. 
    
* Exclusive Resource Dependency Definition
    The possibility to define "resources" the function depend on.
    
*  Serialize execution of "Background  Jobs" which have defined the same dependency of an "Exclusive Resource"
    This makes sure that there are not two different "Background Jobs" which use the same "Exclusive Resource" running at the same time. For example, There might be one SQL job which do SQL maintenance tasks like updating statistics, rebuilds indexes etc.  And then there might be another SQL job which pre-calculate data in cache tables. Running there two at the same time is not desirable, so they both define a specific SQL DB Instance as their "Exclusive Resource". And by doing so, the "Azure Function OS" will ensure that only one of them are running at the same time.
    
* Job progression
   Something similar to `Write-Progress` in PowerShell.
   The progression of a "Background Job" should be shown in the portal and in logs and in AI, etc.
   
* Valid execution time window
    it should be possible to define a "time windows" where the "Background Job" is allowed to be started. Outside this time window it is as the "background job" didn't exist (when it comes to decide which to "Background Job" to run).
    This is to make sure you don't start something which will severely affect the normal operation during business hours.
    
* AI based abnormality detection
   The "Azure function OS" should keep track on the "Background Jobs" and learn what's normal. And then report on stuff which differs from what's normal.
   For each abnormality defected, it should be possible to indicate if it is relevant or not, and why this is (because it happened on a weekend, because it was just after a re-deploy, because this one is run just once a year and you have no clue what's normal and what's not, so just stop trying, etc...)
   
* Uses "Recurrence Rule" (RECUR, RRULE) whenever specifying any form of schedule or recurrence.
     [Recurrence Rule](https://datatracker.ietf.org/doc/html/rfc5545#section-3.3.10) is part of the iCalendar specification and widely used by Microsoft, Google, Apple and many more.
     The Recurrence Rule specification allow for very flexible schedules, where you, for example, can exempt holidays from the schedule.
.
     


### And these limitations
* Somewhat Idempotent, but not really
    A "Background Job" should be coded in such a way that you can execute it any number of times without any problems.
    The definition of "Idempotence" is: 
    > Idempotence: is the property of certain operations in mathematics and computer science whereby they can be applied multiple times without changing the result beyond the initial application.
    
    There is no need for the "Background Job" to generate the exact same result, but it should handle being executed multiple times

* No guaranties on when the "Background Job" will be executed, or if at all
    Execution of  a "Background Job" is done on a "best effort principle". A lot of things could affect time when it is executed, including previous errors in this very same "Background Job", to execution being delayed because of using an "Exclusive Resource", or the time-window for when the "Background Job" can be executed has passed.
    No assumptions should be made on when the "Background Job" is executed.

**Comment by olitomlinson** [comment]
@Bartolomeus-649 

I think the vast majority of what you've described can  be achieved by composing what already exists today in the Durable Functions and Entities API.

> stateless

By shaping your workload into the Eternal Orchestration pattern, you can take advantage of `ContinueAsNew(null)` - no state is maintained between executions.

> singleton

The singleton pattern already exists in Durable Functions. However, It is a bit boilerplate-y as you have to check for statuses etc, but recent fixes have addressed some of the edge cases where the Singleton gaurantee was broken.

>Always active

Once activated an Eternal Orchestration can loop indefinitely via `ContinueAsNew` . You can programmatically pause/resume the  Function using `DurableTimers` to what ever schedule you can represent with code.

> Exception Handling

You are free (and encouraged) to handle your exceptions within the Orchestration logic, and deploy what ever compensatory actions are necessary. Coupling this with `DurableTimers` you are able to code all kinds of back-off strategies.

> Exclusive Resource Dependency Definition

You could enforce this through some logical ownership. I have done similar by using Durable Entities where each entity acts a 'Proxy' to a well-known resource (regardless of if the resource is an ARM resource or an actual API). The proxy owner is determined by First Come First Serve semantics where a potential owner calls a `Lock(string owner)` method on the proxy Entity, and subsequently releases its lock via an `Unlock(string owner)` method.

I'm not an expert, but doing this at the actual ARM level would be a huge challenge I think. You might be able to get somewhere near this using a combination of Managed Identities at the Function App level, but I imagine this wouldn't be granular enough for your use case.

A combination of Orchestration Singletons and appropriate ID conventions for the Orchestration Instance ID would give you that 'single concurrency' semantic that you desire - But lets not forget that Durable Entities and Orchestrations have exclusive access (provided by single concurrency guarantees) to state, so you're already 95% there by choosing the DF framework.

> Job Progression

This can be covered by updating the 'customStatus' property of an Orchestration and is a common use-case.

> Valid execution time window

Once again, using `DurableTimers` and Durable Entity `Reminders` you can execute to a compiled or dynamic schedule - My team did exactly this for a Durable Functions workflow engine that needed to obey timing rules to match our customers working hours.

> AI based anomalie detection

Application Insights (and also Azure Stream Analytics) are the places where you would develop these kind of algorithmic/AI insights from a PaaS perspective. I get the need, but Azure Functions and its runtime simply has no awareness of what "good looks like" for your use-cases.

I would advise getting to a point where your code raises the necessary metrics and health 'signals', which can be tracked and monitored in an appropriate product such as App Insights or ASA or any of the other first-class Azure ML technologies.

> Recurrence Rule

Similar to the execution time window, the API surface is already there to do this. 

What you need is a helper library that can do the heavy lifting of packing and unpacking a complex schedule, and then integrate this with the `Timer` and `Reminders` API of Durable Functions. FYI we rolled our own scheduling instruction set as we  found the Recurrence Rule set to be way to complex, and we could get by with a minimal set of instructions. Of course your mileage may vary here.

However I do agree that Durable Entities supporting Recurrence Rule instruction set for signals, would be a neat addition.

**Comment by Bartolomeus-649** [comment]
@olitomlinson I could have done all this in C back in 1995, but that's not what all this is about, is it?

It's all about packaging and making sure there is no loose ends, making sure it it is easy to use and maintain by the end-developers.

With your arguments, we should close down and abandon Axure functions, we already have App Service and .NET, we can just use that and do it all by ourselves....

**Comment by olitomlinson** [comment]
@Bartolomeus-649 

I’m just saying there are robust APIs and common patterns/concepts in Durable Functions that can achieve many of the things you’re asking for, so it may be worth taking a look at those if you haven’t already done so.

If those APIs are not useful or fall short for your use-cases, then that would be really great feedback!

**Comment by Bartolomeus-649** [comment]
@olitomlinson Have you read what this issue is about???

**Comment by olitomlinson** [comment]
@Bartolomeus-649 

Of course, and as Chris and Seb have both mentioned, there are challenges to this “Start-up” trigger notion.

Until the start-up trigger is available, I’m providing alternatives that may help you.

 I’m also providing guidance on how you can accomplish some of those things with what exists today in the API.

If you do not wish to take that guidance that’s absolutely fine.

**Comment by Bartolomeus-649** [comment]
@olitomlinson, I actually don't understand what you are talking about....did you somewhere switch the subject of this issue from "there's a problem having a function which is always active" to "lets tell this guy, who didn't express he has any problems at all,  how to do stuff with the existing functionality in Azure functions"?

**Comment by cgillum** [comment]
The concept of a background job is interesting, but I feel like the requirements mentioned are too specific. As @olitomlinson suggested, I think it makes more sense for this to be implemented as a layer on top of Durable Functions rather than a built-in feature. I think it's better for the Durable Functions team to instead focus on the primitives needed to make building something like background jobs more manageable. More general purpose features like "auto-start" and "auto-restart" seem like good primitives for us to expose, since implementing those yourselves is pretty non-trivial.

**Comment by Bartolomeus-649** [comment]
@cgillum, which is why I included the line 
> _This list include way more than what's initially needed, the "extra" things are just here to indicate the long term direction "Background Job" should have_

**Comment by Bartolomeus-649** [comment]
I'm sorry I tried to contribute and provide some constructive ideas on this. Won't make the same mistake again.

**Comment by MasterKuat** [comment]
Hi, Up on this subject.
Should be great indeed to have ability to start an orchestration when Azure function is ready (AppLeaseManager ready event ?)

My use case : I have an orchestration with a blob trigger. Sometimes, eventgrid do not reach to deliver event to Azure function. Should be great to have an orchestration that will read dead letter when Azure Function start/restart. This should avoid to setup an external thing (automation or whatever) just to call an http trigger.

**Comment by cgillum** [comment]
@MasterKuat my recommendation would be to use a queue trigger with Event Grid rather than HTTP if you need the delivery of the notification to be reliable.

Closing this issue as not planned since I think this isn't a common requirement, and that there are sufficient workarounds for the mentioned scenarios.
